### ISSUE TITLE
Discuss: Use a Router Pattern For Grant Contributions

### ISSUE THREAD

**Issue by apbendi** [dcgtc/dgrants#20]
Currently, the `GrantRound` contract is architected such that contributions are made to a Grant through the Round. An alternative, and perhaps superior approach, would be to have users make contributions through a `ContributionRouter` contract.

Such a Router would:

* Allow the user to make a contribution directly to the Grant if a Round is not specified
* Allow the user to make a contribution through a Round by specifying a Round address
   * Questions on this point
       * Should the Router then call through to the Round contract to do the transfer?
       * OR should the Router simply verify that the Round is active, and that the donation token is correct, then do the transfer itself and emit an event with the Round contract address included and indexed?

Advantages to this approach include:

* Enabling trackable/indexed contributions outside the context of a active Round
* Reducing the number of times a user needs to `approve` a contract that can do the appropriate `transferFrom`

Possible downsides:

* Depending on how it's architected, it might require each Round to whitelist a Router (or Router**s**?) that is approved to process contributions

Other questions:

* Should this routing functionality actually be built into the `GrantRegistry` directly, as opposed to in a separate Router contract? What are the tradeoffs for each approach?
* Should matching contributions also be sent through a Router?
* If we wanted to implement a feature that allowed users to donate in any Token, then swapped into the Token required by the `GrantRound` before making the contribution, how would the Router approach impact this? Could/should swapping be implemented in the Router, or in another contract wrapping the Router?
* Would the Router also be the right place to enable bulk contributions?

**Comment by mds1** [comment]
I like the idea of making contributions through a static router/contract—instead of through the `GrantRound` contract—because it will save a lot of money in approval fees in the long run, plus improve UX.

Here's a quick run-down of some approaches and tradeoffs I can think of as to how this should be implemented, where we assume the following struct exists:

```solidity
struct Donation {
    uint96 grantId; // grant ID
    address token; // address of the token to donate
    uint256 amount; // amount of tokens to donate
}
```

## Standalone Router

How it works
- `GrantRegistry` defines a list of grants
- User approves `ContributionRouter` to spend their tokens
- `ContributionRouter` contract contains a `bulkDonate` method with the function signature `bulkDonate(address round, Donation[] donations)`, where `round` identifies which round you are donating for
- `ContributionRouter.bulkDonate()` loops through each `donation` and for each, gets the grant's `payee` address from the `GrantRegistry`, transfers tokens to that address, and emits an event

Trade-offs
- ✅  One time approval of Router
- ✅  Easy to add new functionality (e.g. token swaps) by deploying a new router and requiring new approvals
- ❌  CALL to `GrantRegistry` in order to read it's storage adds 2600 gas per bulk donation
- 🟡  Does this mean you have to use the correct Router, and if you don't your contribution is not counted? For example, say a second router is deployed to add new functionality. Are contributions through the old one no longer counted?

## Donate Through Registry

How it works
- `GrantRegistry` defines a list of grants
- User approves `GrantRegistry` to spend their tokens
- `GrantRegistry` contract contains a `bulkDonate` method with the function signature `bulkDonate(address round, Donation[] donations)`, where `round` identifies which round you are donating for
- `GrantRegistry.bulkDonate()` loops through each `donation` and for each, gets the grant's `payee` address from the `GrantRegistry`, transfers tokens to that address, and emits an event

Trade-offs
- ✅  One time approval of GrantRegistry
- ❌ Hard to add new functionality (e.g. token swaps)
- ✅  No CALL to `GrantRegistry` needed to read it's storage
- ✅  Requiring donations through the registry makes it easy to always find all past contribution logs

## Donate Through Round

How it Works
- Similar to above, but instead the `bulkDonate` method exists on the round you are donating for

Trade-offs
- ❌  Have to re-approve each individual grant round contract
- ✅ Easy to add new functionality by changing the `GrantRoundImplementation` used by the `GrantRoundFactory`
- ❌  CALL to `GrantRegistry` in order to read it's storage adds 2600 gas per bulk donation
- ✅ Easy to find all past contribution logs by funding all logs for all deployed rounds

## Other Notes

1. For every single donation, in all of the above cases, we must do a lookup from grant ID to `payee` address, which always requires an SLOAD costing 2100 gas per grant. This adds up quickly! It costs 210,000 gas just to read the addresses of where to send payment for 100 grants. We should instead consider making the `payee` address the unique identifier and not allowing two grants to share the same `payee` address (obvious attack vector here is to create a grant and use someone else's multisig address, so they can't use it, so need some thought around how to do this while still allowing contracts as the receiving address)

1. One solution to repetitive approvals is to require that the donation tokens support `permit`, so we can ask for a signature and do the approvals in the same transaction. A `permit` each round is still more expensive than one approval, but it does allow the "Donate Through Round" option to have better UX

**Comment by apbendi** [comment]
Great write up @mds1! Before I comment on your options, I think there's (at least) one more option worth considering. Would love feedback on it:

### Donate Through Round With Router

How it works:

* The `GrantRound` contract maintains it's single donation method
* The `ContributionRouter` contract adds a bulk contribution method that loops over the specified contributions and calls the `GrantRound` contract for each

Trade-offs

* ✅ One time approval of Router
* `GrantRound` must whitelist router at deployment (because the Router will be the `msg.sender` of the contribution, we couldn't trust any arbitrary router to be honest about who was making the contribution)
* ✅ Fairly easy to add new functionality (deploy a new Router + update the Dapp to whitelist said router when deploying new `GrantRound` instances; existing `GrantRound`'s can't update
* ❌ CALL to GrantRegistry in order to read it's storage adds 2600 gas per bulk donation
* ❌ CALL to `GrantRound` donate method also adds overhead

---

> We should instead consider making the payee address the unique identifier and not allowing two grants to share the same payee address

I hear you on this, but I definitely don't like the idea of making the payee address the unique identifier. In addition to the abuse vector you mentioned, it's also problematic that we want to enable Grant owners to change the payee address, which would change the unique identifier, which is a big no-no.

If we're very concerned about this gas overhead from the lookup, it might be an argument in favor of deploying a Contract to represent the Grant, as opposed to just a struct entry. In that case, the Grant contract address could be the identifier, and the contract could also receive the payment, with the owner simply having the privilege to sweep the funds.

**Comment by mds1** [comment]
> Donate Through Round With Router

Ah interesting. I think you could still have the bulk donation method on the `GrantRound` to save on the number of CALLs, and the router can just be responsible for swaps and managing e.g. donations to multiple rounds in a single transaction. This is another good option!

------------------------

> I hear you on this, but I definitely don't like the idea of making the payee address the unique identifier. In addition to the abuse vector you mentioned, it's also problematic that we want to enable Grant owners to change the payee address, which would change the unique identifier, which is a big no-no.
>
> If we're very concerned about this gas overhead from the lookup, it might be an argument in favor of deploying a Contract to represent the Grant, as opposed to just a struct entry. In that case, the Grant contract address could be the identifier, and the contract could also receive the payment, with the owner simply having the privilege to sweep the funds.

IMO the gas overhead is pretty steep and worth mitigating.

My thinking was we'd do this pretty much the way you described: when a new grant is created we deploy a DSProxy owned by `owner`, so only `owner` can withdraw funds from it. This way the `payee` address is constant, unique address, but you can change the `owner` who can withdraw funds.

It costs about [600k gas](https://etherscan.io/tx/0x1baca3a7685b48cd1d7fd70ecc307e59b8f9163bc3d0983c568f15097630e9d5#ContentPlaceHolder1_collapseContent) to deploy a DSProxy, but they are deployed as full contracts instead of minimal proxies. We can probably make it a bit cheaper with minimal proxies, and could run the numbers on how much cheaper + how many DSProxy transactions you need to break even on gas cost

**Comment by phutchins** [comment]
Something to consider here is a requirement that we clarified today. We would like to allow a donation to a single Grant to be counted for multiple Grant Rounds. It sounds like some of the solutions mentioned here would make it possible to do this.

**Comment by apbendi** [comment]
> Something to consider here is a requirement that we clarified today. We would like to allow a donation to a single Grant to be counted for multiple Grant Rounds. It sounds like some of the solutions mentioned here would make it possible to do this.

Hmm I actually missed or misunderstood this discussion, I think. As it stands I don't think the architecture allows this, and it's not clear how it could. Perhaps one approach is to have the canonical router take a list of qualifying `GrantRounds`, then emit events for each upon donation. What do you think @mds1?

**Comment by mds1** [comment]
Hmm, so let's spec this out a bit more first—if I'm doing a bulk donation, do we want *every* donation to be counted towards the same _n_ rounds, or is on a per-donation basis? In other words, which do we want:

```solidity
// Option 1
struct Donation {
  uint96 grantId;
  uint256 donationAmount;
}

function donate(address[] roundAddresses, Donation[] donations) external {
    // every donation is counted towards every round in roundAddresses
}

// Option 2
struct Donation {
  uint96 grantId;
  uint256 donationAmount;
  address[] roundAddresses;
}

function donate(Donation[] donations) external {
    // every donation has it's own array of rounds to count towards
}
```

**Comment by mds1** [comment]
I think in either case, one option is to add a `roundAddresses` array emitted with the event, but then I don't think you can index/filter on a given round address. 

Another option is to emit 1 event per round per donation, but this feels a bit wasteful in terms of gas

**Comment by apbendi** [comment]
I think we'd want Option 2. I think we'd probably want one event per donation. While a bit costly, I think it'd make event indexing/filtering on the frontend *way* easier.

**Comment by mds1** [comment]
Hmm, it does feel a bit gross to have to emit duplicate events, so I'm wondering—do we even need to specify the round addresses at all? In other words, is there any reason you would want to exclude your contribution from an ongoing round if it's eligible for that round? Because an alternative approach is just leave it up to the round operator to filter for events emitted between their `startTime` and `endTime`, and exclude events from contributions to grants that are not part of their round.

**Comment by thelostone-mc** [comment]
I can clarify a bit more here with an example !
This is what to on Gitcoin Grants at this moment.

If a `Grant X` is part of `2 Grant Rounds` (dApp and Asian Round) ->  any contribution made to `Grant X` would be counted for both `Grant Rounds`.

Reason for doing this from @owocki: 
> the more i think about it; the more i am of the belief that Grant X can be in N matching rounds, not 1 is an important design goal.
why is it important?  well, because the network effect of gitcoin/QF is that everyone else is playing the game, and there’s compounding value.  eg every user makes it better for every other user.    if grants matching amounts can’t stack, then we blunt our greatest advantage over any other grants protocol - our network effect.

While this complicates our current architecture -> it does make sense on why.
I do feel this increases attack vector as now if an address is identified as sybil -> every GrantRound would have to keep track of this. In our current architecture -> since donations goes via a GrantRound -> we could delegate that responsibility to the Round Creater (we still are open to attack but it's got lesser surface area )

======

This means

- ❌ contributing via GrantRound is not an option 
- ❌ contributing via GrantRound -> ContributionRouter
- ✅ Standalone ContributionRouter
- ✅ ContributionRouter -> GrantRegistry 

======

This does complicate things further cause consider the same example where
- dApp Round is run in DAI
- Asian Round is run in DOGE

The swap logic wouldn't make sense as a contribution is being made against the grant and not a grant round.
This means we'd have to capture the DAI value at the time of donation (fire that via an event) -> and that info would be used to determine the distribution  

If I contribute in 10 BADGER -> the event fired would have to record the USD value (70$) so that this amount can be taken while determining the distribution matching

**Comment by mds1** [comment]
Just to be clear, my understanding is that we have finalized "a single contribution must be able to count towards matching for multiple rounds ongoing at the same time" as a system requirement. So the rest of this comment (and the rest of discussion in this issue, unless otherwise stated?) will assume that

------------------------

> do we even need to specify the round addresses at all? In other words, is there any reason you would want to exclude your contribution from an ongoing round if it's eligible for that round? Because an alternative approach is just leave it up to the round operator, when doing matching calculations, to filter for events emitted between their startTime and endTime, and exclude events from contributions to grants that are not part of their round.

------------------------

I still feel like this is the best way to handle the "what rounds are you contributing to" question. Are there objections here?

> I do feel this increases attack vector as now if an address is identified as sybil -> every GrantRound would have to keep track of this

This seems ok. I think Grant Rounds should be able choose to use their own criteria for determining if an address is a sybil, so just because GrantRound1 identifies some users as sybil doesn't mean GrantRound2 will agree. I think this is a problem that we should think about at a different layer, but not part of scope for this "determine contribution Router pattern" issue

------------------------

> This means revisiting the flows we discussed above (for contributing to a grant and having it accounted for in rounds):
> 
> ❌ contributing via GrantRound is not an option
> ❌ contributing via GrantRound -> ContributionRouter
> ✅ Standalone ContributionRouter
> ✅ ContributionRouter -> GrantRegistry
> ✅ contributing via.GrantRegistry (this seems the most beneficial )

Agreed, this is a bit nice in that it helps us remove some options. I tend to agree that contributing via the `GrantRegistry` is optimal (mainly for gas reasons), but we should probably (1) think about what those updated might be in reality, and (2) discuss a way to ensure we aren't locked in to contribution logic forever in case it needs to be updated

Another benefit of using the Registry is that is simplifies communications/authority decisions. Who gets to decide what is the "official" router that all rounds should be using? And how do you communicate to all entities when a new router is deployed? An example: what if (1) GrantRound1 starts with Router1, then (2) Router2 is deployed, then (3) GrantRound2 starts with Router2—it now is difficult to ensure my contribution is counted for matching in both GrantRound1 and GrantRound2 because they use different routers

------------------------

> This does complicate things further cause consider the same example where
> 
> - dApp Round is run in DAI
> - Asian Round is run in DOGE
> 
> The swap logic wouldn't make sense as a contribution is being made against the grant and not a grant round.
> This means we'd have to capture the DAI value at the time of donation (fire that via an event) -> and that info would be used to determine the distribution
> 
> If I contribute in 10 BADGER -> the event fired would have to record the USD value (70$) so that this amount can be taken while determining the distribution matching


Agreed this is probably the trickiest part. I think there's a few options here:
1. GrantRegistry chooses the donation token and every Round has to use that token
2. GrantRegistry has a list of approved tokens, and an event is emitted containing the price-per-token for each token at the time of donation
3. Back to the original model of "grants receive donations in any token" and price conversions are left to off-chain entities 

1 and 2 make the GrantRegistry less neutral/permissionless, so I'm not a fan of them. I think option 3 is the best, and leaving it to the round operator to get off-chain price data isn't any harder than off-chain match calculations, so this seems ok to me. There may be other options too, these are just the first that come to mind

**Comment by apbendi** [comment]
> do we even need to specify the round addresses at all? In other words, is there any reason you would want to exclude your contribution from an ongoing round if it's eligible for that round? Because an alternative approach is just leave it up to the round operator, when doing matching calculations, to filter for events emitted between their startTime and endTime, and exclude events from contributions to grants that are not part of their round.

> I think option 3 is the best, and leaving it to the round operator to get off-chain price data isn't any harder than off-chain match calculations, so this seems ok to me. There may be other options too, these are just the first that come to mind

The issue I see here is, by moving more and more of this off chain, and into the hands of the entities that choose to run rounds, we're simply recreating a centralized system. At best, we'd have a universal GrantRegistry, and a set of open source tools for centralized entities to run their own Rounds. Is that the direction we want to go?

**Comment by mds1** [comment]
I don't think moving price conversions off-chain is too big of a centralization issue though for two reasons:
- Like you've mentioned, doing secure on-chain price conversions can be tricky, especially for low liquidity tokens, so you're arguably better off doing it off-chain for security reasons
- Price conversions are only needed for computing match amounts, and that is currently off-chain anyway, so this is just an additional component in the "verify match calculations" step

**Comment by apbendi** [comment]
> doing secure on-chain price conversions can be tricky, especially for low liquidity tokens, so you're arguably better off doing it off-chain for security reasons

By swapping, we're not doing a price conversion, we're *actually* converting the token, which is a *feature* not a downside, because it leaves no room for arguing over or attempting to manipulate the value. Calculating the "price" of a low liquidity token off-chain opens the system up to manipulation and dispute. Even for high-liquidity token, the choice of off-chain data source for price could have non-trivial impacts on the matching results. Furthermore, if that off-chain data source is centralized, it's also non-deterministic (imagine the source is the coingecko api, and coingecko updates its historical data or even just goes offline). More fundamentally, using a centralized source for price history cedes a chunk of security to the entity chosen.

> Price conversions are only needed for computing match amounts, and that is currently off-chain anyway, so this is just an additional component in the "verify match calculations" step.

Reiterating the above, it's "just an extra step," but it's a step that opens the system to attack, reduces reproducibility, and dramatically increases the centralization.

Furthermore, we're not *just* talking about moving price conversions off-chain. You said:

> an alternative approach is just leave it up to the round operator, when doing matching calculations, to filter for events emitted between their startTime and endTime, and exclude events from contributions to grants that are not part of their round

This hands a lot more responsibility and discretion to the operator, which we're already handing a lot to. It also forecloses on any clear path for *eventually* introducing a decentralized, on-chain system for handling these concerns.

I think my previous characterization was correct: if we go this route, we're starting to drift toward is an open source implementation of centralized Grants with a shared registry.

**Comment by mds1** [comment]
> we're actually converting the token, which is a feature not a downside, because it leaves no room for arguing over or attempting to manipulate the value

I think quoted part is where our disconnect is. I totally understand and agree with your points, but was coming from the perspective that converting the token on-chain is no longer a solution because of this problem that @thelostone-mc mentioned above:

> This does complicate things further cause consider the same example where
> 
> - dApp Round is run in DAI
> - Asian Round is run in DOGE
> The swap logic wouldn't make sense as a contribution is being made against the grant and not a grant round.

I agree that if all ongoing rounds used the same token, then swapping would be the way to go and it solves the price issues. But if a contribution is eligible for multiple ongoing rounds that use different tokens, a swap doesn't help with that oracle problem and it needs to be solved another way. Ways I can think of to solve it are:

1. Leave the issue of determining price to the round operator as part of matching calculations. With this approach, a contribution can be automatically "entered" into every ongoing round, where entered is at the discretion of the round operator who filters for all event logs between the round start and end times
2. Require contributions to specify which rounds they are eligible for, and for each, use TBD oracle solution(s) to determine an on-chain price and emit that as an event. The challenges with this are (1) getting oracle prices might be gas heavy, and (2) generalizing a way to get prices that works for most/all tokens is a challenege

**Comment by apbendi** [comment]
> but was coming from the perspective that converting the token on-chain is no longer a solution because of this problem that @thelostone-mc mentioned above:

Got it! Fair enough. My response was intended as pushback on that perspective as something now taken for granted. The compromises we'd have to make are pretty big, so is it definitely worth making them?

It seems to me that the two confounding features that cause this issue are:

1. Allowing a contribution to count for multiple rounds
2. Allowing different Round's to accept arbitrary tokens.

If we're sure we want no. 1, then perhaps we should reconsider no. 2, rather than making all the other compromises I enumerated in order to support both.

In particular, I could imagine the RoundFactory we use enforcing the token set for the Round. The Router could swap to the token specified by the Factory. Then any given contribution could, at a minimum, be safely applied to every Round that was deployed using a particular Factory. Lots of details to work out here, but (I think) at least technically feasible.

In this scenario, another entity could still choose to deploy their own Factory and create a separate universe of Rounds that use a different donation token, though the Registry and the Grants there would remain shared. My guess is that this would be uncommon, as a given "Round Universe" would accrue its own set of network effects.

One cool possibility I'll throw out here is that GTC could be chosen as the default token for the default factory, possibly helping establish it as a schelling point currency for public goods funding in the ecosystem.

**Comment by gdixon** [comment]
From what I can garner by playing with the CLR calc - it shouldn't matter what currency we feed in or match against - all that matters is that the contributions are representative of the initial input.

If you have all contributions denominated in ETH and match with DODGE, we would get the same result if we converted all of those ETH contributions to DODGE (using the same conversion rate) or if we just ran the calculation with the ETH input. The number of contributions and their relative weight when compared to each other would be equal and the final match would be the same.

I think the safest way to handle this is to convert everything internally to a common currency - I've fleshed this thought out a little bit more on this comment: https://github.com/dcgtc/dgrants/issues/24#issuecomment-884961970

I'm not sure, but I think this may complicate things when producing the prediction curve? If we're not at the final figures, would this still work? @thelostone-mc

**Comment by phutchins** [comment]
(i hope all of this makes sense as there were a lot of different issues to comment on...)

> so I'm wondering—do we even need to specify the round addresses at all? In other words, is there any reason you would want to exclude your contribution from an ongoing round if it's eligible for that round? Because an alternative approach is just leave it up to the round operator, when doing matching calculations, to filter for events emitted between their startTime and endTime, and exclude events from contributions to grants that are not part of their round.

- One might want to exclude their grant from a round but I don’t have any good reasons why currently
- This could be something we move out of scope and just say for the time being, a grant cannot be excluded from any round

> If I contribute in 10 BADGER -> the event fired would have to record the USD value (70$) so that this amount can be taken while determining the distribution matching

- This could done at the convenience layer and could be calculated after the fact given a good API for historical crypto values
- The convenience layer could also do it near real time as events come in, record the values, then make that available for consumption via public API (yes, room for decentralized automation here but this could just be a stepping stone)

> I still feel like this is the best way to handle the "what rounds are you contributing to" question. Are there objections here?- I think this seems right to me
> This seems ok. I think Grant Rounds should be able choose to use their own criteria for determining if an address is a sybil, so just because GrantRound1 identifies some users as sybil doesn't mean GrantRound2 will agree. I think this is a problem that we should think about at a different layer, and not part of scope for this "determine contribution Router pattern" issue

- In my mind I think we control this at the entity level but still give the GrantRound creator simple choices. If you create a grant on the Gitcoin Grants explorer (the explorer run by Gitcoin), it has the option to use Gitcoin sybil defense, potentially some other sybil defense API (will be part of the protocol), or do it yourself sybil defense (which could mean none).
- Most importantly I think that it should be made clear upon creation of the GrantRound what level of sybil filtering is intended to be done while calculating matching (more discussion to be had here tho)

> Using different routers because of upgrade/update?
- The calculations are done using the data/events provided from the router so we should simply be able to update the payout calculation script/mechanism to account for the small differences in how the routers flag contributions

> Agreed this is probably the trickiest part. I think there's a few options here:
> 	1.	GrantRegistry chooses the donation token and every Round has to use that token
> 	2.	GrantRegistry has a list of approved tokens, and an event is emitted containing the price-per-token for each token at the time of donation
> 	3.	Back to the original model of "grants receive donations in any token" and price conversions are left to off-chain entities
> 1 and 2 make the GrantRegistry less neutral/permissionless, so I'm not a fan of them. I think option 3 is the best, and leaving it to the round operator to get off-chain price data isn't any harder than off-chain match calculations, so this seems ok to me. There may be other options too, these are just the first that come to mind

- I Would prefer that we accept more tokens initially at the cost of doing more off chain at least in the beginning. We can work our way toward more decentralization/on-chain in this area down the line.
- If we abstract that out so that we're only accepting GTC or something while we build out the router module that makes sense to me as well

> Off chain token value pegging vs on chain & donations to multiple rounds in different tokens

- For the current scope, lets calculate off chain and make it modular so that we could move it on chain
- I would love to see a more in depth comparison of the two approaches focusing on cost
- This is another area where I think that an L2 could be beneficial
- We should make a list of areas where L2’s would be helpful and create an initiative to push on them all in parallel (one could learn from the other)
- I do think that down the road on chain capture of value at a given time will be the right way to go but in the short to maybe mid term I think we build something into the convenience layer (centralized API that can be run by any entity that wants to run one) that monitors events and captures values at each time of contribution. It seems like we have to peg value to something consistent across the board which seems like USD for the moment. 
- The important part in my mind is to give easy to use transparent tools through a single user interface (or repository when necessary) enabling verifiable results 

> an alternative approach is just leave it up to the round operator

- I’m of the feeling that we should leave as little to the operator as possible (especially because they are not expected to have any control over the round after creation. This being the case, we are putting a lot more control in their hands. Yes the data can be validated but it is very difficult and time consuming to do that which makes it unlikely for anyone to do it.)

> Allowing different rounds to accept arbitrary tokens

- I think we can start with allowing different rounds to accept any of a predefined set of tokens
- We can if needed set the round/grants up to accept only one token and leave a modular slot in place where a router/token converter would go which we can solve later

> GTC for chosen default token
- I like this idea

**Comment by mds1** [comment]
I want to lay out a rough spec for a potential implementation that @gdixon and I talked about earlier today and can be inferred from his comments here and in #24. The implementation would work like this, where we start by defining a data type:

```solidity
// Define inputs required for each contribution in a struct
struct Donation {
  uint96 grantId; // grantId
  address token; // address of token user is donating in
  uint256 amount; // amount of that token
  address[] rounds; // rounds they want the contribution to count for
}
```

Now:

1. Donations occur through the `GrantRegistry` (or a Router if we think that's better for some reason, but as of now Registry seems to be the better option since it saves a lot of gas, and we're about to get a little gas guzzly)
2. A user calls a method on `GrantRegistry` with the following function signature: `function donate(Donation[] donations)`
3. For each `donation` in `donations`, we:
    1. Verify this is a valid contribution for each `round` specified
    2. Swap from `token` to GTC (or ETH, this is discussed more below) with Uniswap V3. We receive `gtcAmount` tokens
    3. Send the GTC to the `payee` specified by `grantId`
    4. Emit `DonationSent(grantId, msg.sender, gtcAmount, tokenAmount)`, where `tokenAmount` is arguably not required, but may be helpful for UX, as this is the donation amount specified by the user when checking out

Pros:
1. All donations are received in a single token, which solves the token price oracle problem for this use case
2. Match calculations are done with GTC amounts as the input, and the output is "percentage of match pool each grant receives". This means it does not matter what donation token a round uses for match payouts
3. Minimizes centralization risks

Cons:
1. Forces grant owners to receive GTC (but since they're only receiving one token, they can easily swap out of it at the end of a round)
2. Swaps make contributions pricier

Notes:
1. The above implementation can clearly be optimized in a few ways:
    1. Batch all swaps from `token` to GTC, instead of doing them in a loop—it's just laid out this way for simplicity
    2. On-chain verification (and storage?) that each donation is a valid contribution for each `round` specified is gas guzzly, so perhaps move this off-chain where round operator filters event logs for valid contributions
2. GTC is a nice choice of receiving token because it means grant owners can now participate in governance (including in the match payouts vote). But using ETH instead might make the GrantRegistry more credibly neutral
3. The `msg.sender` emitted in `DonationSent` is intended to be the donor, however this is not accurate true as some contributions/wallets (such as Argent) will use relayers to send the donation. You'd need to get the corresponding `Transfer` log to find the true donation address

(@gdixon—please correct me if I misrepresented anything here!)

**Comment by mds1** [comment]
Replaced by https://github.com/dcgtc/dgrants/issues/40
